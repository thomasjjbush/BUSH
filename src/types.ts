import { ReactElement, UIEvent } from 'react';

export interface AspectRatioProps {
    children?: ReactElement | ReactElement[] | string;
    ratio: AspectRatios;
}

export enum AspectRatios {
    '1:1' = 100,
    '3:2' = 66.66,
    '4:3' = 75,
    '8:5' = 62.5,
    '16:9' = 56.25,
    '32:9' = 28.125,
}

export interface ButtonProps {
    ariaLabel?: string;
    children: Children;
    design?: Buttons;
    onClick: () => void;
}

export enum Buttons {
    PRIMARY,
    SECONDARY,
}

export interface CarouselProps {
    children: ReactElement[];
    color?: string;
    itemsToDisplay?: number;
    margin?: number;
    numberOfItems: number;
    startingIndex?: number;
}

type Children = ReactElement | ReactElement[] | string | string[] | (ReactElement | string)[];

export interface Client {
    logo: {
        url: string;
    };
    name: string;
    primaryColor: string;
    secondaryColor?: string;
    slug: string;
    url: string;
}

export enum ClientsActions {
    LOADED = 'clients/LOADED',
}

export interface ClientProps {
    alt?: string;
    background?: boolean;
    border?: string;
    logo: string;
    size: number;
    url?: string;
    width: number;
}

type Collection<T> = { items: T[]; total?: number };

export interface Employment {
    companyName: string;
    endDate?: string;
    location: {
        lat: string;
        lon: string;
    };
    responsibilities: string;
    startDate: string;
    title: string;
    url: string;
}

export enum EmploymentActions {
    LOADED = 'employment/LOADED',
}

export interface ErrorProps {
    button?: ReactElement;
    code: number;
}

export interface Errors {
    [key: number]: string;
}

type Gallery<T> = Collection<T> & { isLandScape: boolean };

export type GetState = () => Store;

export interface HiddenTextProps {
    children: string;
    lines: number;
}

export interface IconProps {
    color?: string;
    disabled?: boolean;
    icon: Icons | string;
    onClick?: () => void;
    size?: number;
    testID?: string;
}

export enum Icons {
    CONTRAST = 'e9d5',
    ERROR = 'ea07',
    GITHUB = 'eab0',
    INFO = 'ea0c',
    LEFT = 'ea44',
    LINKEDIN = 'eac9',
    RIGHT = 'ea42',
    WHATSAPP = 'ea93',
}

export interface ImageProps {
    alt?: string;
    height?: number;
    quality?: number;
    size: number;
    src: string;
    width?: number | string;
}

interface Media {
    description?: string;
    name: string;
    url: string;
}

export enum Platform {
    DESKTOP,
    MOBILE,
    TABLET,
}

export interface Project {
    client: Client;
    description: string;
    gallery?: Gallery<Media>;
    hero: Media;
    name: string;
    primaryTag?: Tag;
    responsibilities: Collection<Responsibility>;
    slug: string;
    tags: Collection<Tag>;
    thumbnail: Media;
    video: Media;
    year: number;
}

export enum ProjectActions {
    CLEARED = 'project/CLEARED',
    ERROR = 'project/ERROR',
    LOADED_PROJECT = 'project/LOADED',
}

export interface ProjectState {
    error?: number;
    item?: Project;
    sameClient?: Partial<Project>[];
    sameTag?: Partial<Project>[];
}

export enum ProjectsActions {
    LOADED = 'projects/LOADED',
    LOADED_MORE = 'projects/LOADED_MORE',
    LOADING_MORE = 'projects/LOADING_MORE',
}

export interface ProjectsState extends Collection<Project> {
    loading?: boolean;
}

export interface Responsibility {
    description: string;
    icon: string;
    name: string;
}

export interface ScrollerProps {
    children: Children;
    loading?: boolean;
    loadMore?: () => void;
    numberOfItems?: number;
    onScroll?: (e: UIEvent<HTMLDivElement>) => void;
    testID?: string;
    totalItems?: number;
}

export interface Store {
    clients: Client[];
    employment: Employment[];
    project: ProjectState;
    projects: ProjectsState;
    tags: Tag[];
}

export type StyledProps<T = {}> = { theme: Theme } & T;

export interface Tag {
    name: string;
    slug: string;
}

export enum TagActions {
    LOADED = 'tag/LOADED',
}

export interface TagProps {
    tags: Tag[];
}

export interface Theme {
    breakpoints: {
        mobile: number;
        tabletLandscape: number;
        tabletPortrait: number;
    };
    colors: {
        highlight: string;
        background: string;
        border: string;
        brand: string;
        offset: string;
        text: string;
    };
    images: {
        logo: {
            pink: string;
            white: string;
        };
        loading: string;
    };
    isWebpSupported?: boolean;
}

export interface ThumbnailProps {
    client?: ReactElement;
    color: string;
    name: string;
    ratio?: AspectRatios;
    url: string;
}

declare global {
    interface Window extends Window {
        __REDUX_DEVTOOLS_EXTENSION__?: Function;
        __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: Function;
    }
}
