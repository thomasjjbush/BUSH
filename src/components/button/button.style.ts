import styled from 'styled-components';

export const StyledButton = styled.button`
    align-items: center;
    align-self: flex-start;
    appearance: none;
    border-radius: 5px;
    cursor: pointer;
    display: flex;
    font-size: 18px;
    font-weight: bold;
    justify-content-space-between;
    margin: 20px 0 0 0;
    padding: 10px 20px;

    img {
        margin-left: 10px;
    }
`;

export const StyledPrimaryButton = styled(StyledButton)`
    background-color: ${(props): string => props.theme.colors.brand};
    border: 0;
    color: white;
`;

export const StyledSecondaryButton = styled(StyledButton)`
    background-color: transparent;
    border: solid 2px ${(props): string => props.theme.colors.brand};
    color: ${(props): string => props.theme.colors.text};
`;
