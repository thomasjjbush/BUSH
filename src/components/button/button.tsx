import React, { FunctionComponent, ReactElement } from 'react';
import { ButtonProps as Props, Buttons } from '../../types';
import { StyledPrimaryButton, StyledSecondaryButton } from './button.style';

export const Button: FunctionComponent<Props> = (props: Props): ReactElement => {
    if (!props.onClick) return null;
    return props.design === Buttons.PRIMARY ? (
        <StyledPrimaryButton {...props} type="button">
            {props.children}
        </StyledPrimaryButton>
    ) : (
        <StyledSecondaryButton {...props} type="button">
            {props.children}
        </StyledSecondaryButton>
    );
};

Button.displayName = 'Button';
