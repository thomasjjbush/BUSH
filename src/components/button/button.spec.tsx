import React from 'react';
import { shallow } from 'enzyme';
import { Button } from './button';
import { StyledPrimaryButton, StyledSecondaryButton } from './button.style';
import { ButtonProps, Buttons } from '../../types';

describe('Button', () => {
    let props: ButtonProps;

    beforeEach(() => {
        props = {
            children: 'Label',
            design: Buttons.PRIMARY,
            onClick: jest.fn(),
        };
    });

    it('should render null if no onClick cb is provided', () => {
        const wrapper = shallow(<Button {...props} onClick={undefined} />);
        expect(wrapper.isEmptyRender()).toBe(true);
    });

    it('should render primary button correctly and invoke onClick cb on click', () => {
        const wrapper = shallow(<Button {...props} />);
        const primaryButton = wrapper.find(StyledPrimaryButton);
        primaryButton.simulate('click');

        expect(props.onClick).toHaveBeenCalledTimes(1);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render secondary button correctly and invoke onClick cb on click', () => {
        const wrapper = shallow(<Button {...props} design={Buttons.SECONDARY} />);
        const secondaryButton = wrapper.find(StyledSecondaryButton);
        secondaryButton.simulate('click');

        expect(props.onClick).toHaveBeenCalledTimes(1);
        expect(wrapper).toMatchSnapshot();
    });
});
