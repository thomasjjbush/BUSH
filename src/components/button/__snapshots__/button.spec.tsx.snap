// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render primary button correctly and invoke onClick cb on click 1`] = `
<Styled(styled.button)
  design={0}
  onClick={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  styles="
    align-self: flex-start;
    align-items: center;
    appearance: none;
    border-radius: 5px;
    cursor: pointer;
    display: flex;
    font-size: 18px;
    font-weight: bold;
    padding: 10px 20px;
    margin: 20px 0 0 0;

    img {
        margin-left: 10px;
    }

    background-color: theme.colors.brand;
    border: 0;
    color: white;
"
  type="button"
>
  Label
</Styled(styled.button)>
`;

exports[`Button should render secondary button correctly and invoke onClick cb on click 1`] = `
<Styled(styled.button)
  design={1}
  onClick={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  styles="
    align-self: flex-start;
    align-items: center;
    appearance: none;
    border-radius: 5px;
    cursor: pointer;
    display: flex;
    font-size: 18px;
    font-weight: bold;
    padding: 10px 20px;
    margin: 20px 0 0 0;

    img {
        margin-left: 10px;
    }

    background-color: transparent;
    border: solid 2px theme.colors.brand;
    color: theme.colors.text;
"
  type="button"
>
  Label
</Styled(styled.button)>
`;
